code/datums/components/_component.dm

```diff
@@ -0,0 +1,264 @@
+		qdel(src, TRUE, TRUE)
+		CRASH("Incompatible [type] assigned to a [P]!")
+
+	_JoinParent(P)
+
+/datum/component/proc/_JoinParent()
+	var/datum/P = parent
+	//lazy init the parent's dc list
+	var/list/dc = P.datum_components
+	if(!dc)
+		P.datum_components = dc = list()
+
+	//set up the typecache
+	var/our_type = type
+	for(var/I in _GetInverseTypeList(our_type))
+		var/test = dc[I]
+		if(test)	//already another component of this type here
+			var/list/components_of_type
+			if(!length(test))
+				components_of_type = list(test)
+				dc[I] = components_of_type
+			else
+				components_of_type = test
+			if(I == our_type)	//exact match, take priority
+				var/inserted = FALSE
+				for(var/J in 1 to components_of_type.len)
+					var/datum/component/C = components_of_type[J]
+					if(C.type != our_type) //but not over other exact matches
+						components_of_type.Insert(J, I)
+						inserted = TRUE
+						break
+				if(!inserted)
+					components_of_type += src
+			else	//indirect match, back of the line with ya
+				components_of_type += src
+		else	//only component of this type, no list
+			dc[I] = src
+
+/datum/component/proc/Initialize(...)
+	return
+
+/datum/component/Destroy(force=FALSE, silent=FALSE)
+	enabled = FALSE
+	var/datum/P = parent
+	if(!force)
+		_RemoveFromParent()
+	if(!silent)
+		P.SendSignal(COMSIG_COMPONENT_REMOVING, src)
+	parent = null
+	LAZYCLEARLIST(signal_procs)
+	return ..()
+
+/datum/component/proc/_RemoveFromParent()
+	var/datum/P = parent
+	var/list/dc = P.datum_components
+	for(var/I in _GetInverseTypeList())
+		var/list/components_of_type = dc[I]
+		if(length(components_of_type))	//
+			var/list/subtracted = components_of_type - src
+			if(subtracted.len == 1)	//only 1 guy left
+				dc[I] = subtracted[1]	//make him special
+			else
+				dc[I] = subtracted
+		else	//just us
+			dc -= I
+	if(!dc.len)
+		P.datum_components = null
+
+/datum/component/proc/RegisterSignal(sig_type_or_types, proc_or_callback, override = FALSE)
+	if(QDELETED(src))
+		return
+	var/list/procs = signal_procs
+	if(!procs)
+		procs = list()
+		signal_procs = procs
+
+	var/list/sig_types = islist(sig_type_or_types) ? sig_type_or_types : list(sig_type_or_types)
+	for(var/sig_type in sig_types)
+		if(!override)
+			. = procs[sig_type]
+			if(.)
+				stack_trace("[sig_type] overridden. Use override = TRUE to suppress this warning")
+
+		if(!istype(proc_or_callback, /datum/callback)) //if it wasnt a callback before, it is now
+			proc_or_callback = CALLBACK(src, proc_or_callback)
+		procs[sig_type] = proc_or_callback
+
+	enabled = TRUE
+
+/datum/component/proc/InheritComponent(datum/component/C, i_am_original)
+	return
+
+/datum/component/proc/OnTransfer(datum/new_parent)
+	return
+
+/datum/component/proc/_GetInverseTypeList(our_type = type)
+	#if DM_VERSION >= 513
+	#warning 512 is definitely stable now, remove the old code
+	#endif
+
+	#if DM_VERSION < 512
+	//remove this when we use 512 full time
+	set invisibility = 101
```

Comment:
This is actually needed if you're running less than 512, or components will just straight up not work and runtime like mad.

Reply:
Voodoo